app/services/csv/validate.rb

```
@@ -18,13 +25,53 @@ def process
       CSV.parse(@csv_string, headers: true) { |row| @csv << row }
       raise Exception::InvalidCSV if @csv.empty?
 
-      check_empty_cells
+      @csv.each do |row|
+        row.fields.each { |cell| validate(cell) }
+      end
     end
 
     private
 
-    def check_empty_cells
-      @csv.each { |row| raise Exception::InvalidCSV if row.fields.any?(&:blank?) }
+    def validate(cell)
+      check_empty_cell(cell) && unwanted_html(cell) && check_quotes(cell)
```

error_message = empty_cell?(cell) || unwanted_html?(cell)

raise Exception::InvalidCSV(error) if error

---

unpaired_quotes?

---

Some of the comments here:

1. Naming - make it as consistent and specific as possible.  `check_quotes` isn't super effective.  Check it for what?  `unpaired_quotes?`. Now that is obvious, we want to make sure it doesn't have unpaired quotes.

2.  Lets leverage short-circuit evaluation with `||` to dry up your methods below.  In many of the below methods, you raise an error or return true.  Instead, we can return nil **unless** there is an error. Then return a message.  This will evaluate the statement to true and pass the message into `error_message`.  Then after, we can raise an error if there is a mes